rust/src/tabs.rs

```diff
@@ -57,46 +51,49 @@ impl Tabs {
         use rpc::TabCommand::*;
 
         match cmd {
-            NewTab => Some(Value::String(self.do_new_tab())),
+            NewTab => Some(Value::String(self.do_new_tab(rpc_peer))),
 
             DeleteTab { tab_name } => {
                 self.do_delete_tab(tab_name);
                 None
             },
 
-            Edit { tab_name, edit_command } => self.do_edit(tab_name, edit_command, rpc_peer),
+            Edit { tab_name, edit_command } => self.do_edit(tab_name, edit_command),
         }
     }
 
-    fn do_new_tab(&mut self) -> String {
-        self.new_tab()
+    fn do_new_tab(&mut self, rpc_peer: MainPeer) -> String {
+        self.new_tab(rpc_peer)
     }
 
     fn do_delete_tab(&mut self, tab: &str) {
         self.delete_tab(tab);
     }
 
-    fn do_edit(&mut self, tab: &str, cmd: EditCommand, rpc_peer: MainPeer)
+    fn do_edit(&mut self, tab: &str, cmd: EditCommand)
             -> Option<Value> {
         if let Some(editor) = self.tabs.get(tab) {
-            let tab_ctx = TabCtx {
-                tab: tab.to_string(),
-                kill_ring: self.kill_ring.clone(),
-                rpc_peer: rpc_peer,
-                self_ref: editor.clone(),
-            };
-            editor.lock().unwrap().do_rpc(cmd, tab_ctx)
+            editor.lock().unwrap().do_rpc(cmd)
         } else {
             print_err!("tab not found: {}", tab);
             None
         }
     }
 
-    fn new_tab(&mut self) -> String {
+    fn new_tab(&mut self, rpc_peer: MainPeer) -> String {
         let tabname = self.id_counter.to_string();
         self.id_counter += 1;
-        let editor = Editor::new();
-        self.tabs.insert(tabname.clone(), Arc::new(Mutex::new(editor)));
+        let tab_ctx = TabCtx {
+            tab: tabname.clone(),
+            kill_ring: self.kill_ring.clone(),
+            rpc_peer: rpc_peer,
+            editor_ref: Weak::new(),
+        };
+        let tab_ref = Arc::new(Mutex::new(tab_ctx));
```

Comment:
Since the Editor is the only one using it, couldn't the editor just get the ownership of the `TabCtx` (instead of the Arc Mutex)?